src/dataflow/src/render/join.rs

```diff
@@ -277,4 +277,73 @@ where
             panic!("render_join called on invalid expression.")
         }
     }
+
+    pub fn render_semi_join(
+        &mut self,
+        input: &RelationExpr,
+        keys: &[ScalarExpr],
+        scope: &mut G,
+        worker_index: usize,
+    ) {
+        if let RelationExpr::Join {
+            inputs,
+            equivalences,
+            implementation: expr::JoinImplementation::Semijoin,
+            ..
+        } = input
+        {
+            // Extract the constant to semijoin against from the equivalence classes
+            let temp_storage = RowArena::new();
+            let constant_row = keys
+                .iter()
+                .map(|k| {
+                    let eval_result = equivalences.iter().find_map(|e| {
+                        if e.contains(k) {
+                            return e.iter().find_map(|s| match s {
+                                ScalarExpr::Literal(_, _) => Some(s.eval(&[], &temp_storage)),
+                                _ => None,
+                            });
```

Comment:
I can't read this. I don't know where the `return` returns to. Some comments, or different control flow would help me!

Also, you may want `ScalarExpr`'s function
```
    pub fn as_literal(&self) -> Option<Result<Datum, &EvalError>> {
        if let ScalarExpr::Literal(lit, _column_type) = self {
            Some(lit.as_ref().map(|row| row.unpack_first()))
        } else {
            None
        }
    }
```